applications/nrf5340_audio/src/amod_main.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright(c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include "amod_api.h"
+#include "aobj_api.h"
+#include "data_fifo.h"
+#include "lc3_decoder.h"
+#include <ctype.h>
+#include <nrfx_clock.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <zephyr/kernel.h>
+#include <zephyr/shell/shell.h>
+
+/* Headers below only for testing */
+#include "amod_api_private.h"
+#include "lc3_decoder_private.h"
+
+#define MODULE_DATA_SIZE (4096)
+#define LC3_DECODER_DATA_SIZE AMOD_DATA_OUT_SET_SIZE(FRAME_SIZE_BYTES)
+#define LC3_DECODER_IN_MESSAGE_NUM (2)
+#define LC3_DECODER_OUT_MESSAGE_NUM (2)
+#define LC3_DECODER_DATA_NUM (4)
+
+#define LC3_DECODER_THREAD_STACK_SIZE (4096)
+#define LC3_DECODER_THREAD_PRIORITY (2)
+#define LC3_DECODER_CHANNELS (2)
+#define LEFT_CHANNEL BIT(0)
+#define RIGHT_CHANNEL BIT(1)
+
+char *lc3_decoder_instance_name = "LC3 Decoder inst 1";
+char private_mem_static[WB_UP(MODULE_DATA_SIZE)];
+char in_msg_mem_static[AMOD_IN_MSG_SET_SIZE(LC3_DECODER_IN_MESSAGE_NUM)];
+char out_msg_mem_static[AMOD_OUT_MSG_SET_SIZE(LC3_DECODER_OUT_MESSAGE_NUM)];
+char data_mem_static[AMOD_DATA_BUF_SET_SIZE(WB_UP(FRAME_SIZE_BYTES), LC3_DECODER_CHANNELS,
+					    LC3_DECODER_DATA_NUM)];
+
+K_THREAD_STACK_DEFINE(lc3_decoder_thread_stack, LC3_DECODER_THREAD_STACK_SIZE);
+
+struct lc3_decoder_configuration lc3_dec_inst_1_config = { .sample_rate = 48000,
+							   .bit_depth = 16,
+							   .duration_us = 10000,
+							   .number_channels = LC3_DECODER_CHANNELS,
+							   .channel_map = (LEFT_CHANNEL ||
+									   RIGHT_CHANNEL) };
+
+/**
+ * @brief Static allocate memory for the LC3 decoder module.
+ *
+ */
+static int static_allocate_memory(struct amod_handle *handle, char **in_msg_mem, char **out_msg_mem,
+				  char **data_mem)
+{
+	*handle = &private_mem_static[0];
+	*in_msg_mem = &in_msg_mem_static[0];
+	*out_msg_mem = &out_msg_mem_static[0];
+	*data_mem = &data_mem_static[0];
+
+	return 0;
+}
+
+/**
+ * @brief Dynamic allocate memory for the LC3 decoder module.
+ *
+ */
+static int dynamic_allocate_memory(struct amod_parameters *parameters,
+				   struct amod_configuration mod_config, struct amod_handle *handle,
+				   char **in_msg_mem, char **out_msg_mem, char **data_mem)
+{
+	int ret;
+	int handle_size, in_msg_size, out_msg_size, data_size;
+	struct _amod_parameters *params = (struct _amod_parameters *)parameters;
+
+	handle_size = amod_query_resource(params, mod_config);
+	if (handle_size < 0) {
+		printf("Querry resource FAILED , ret %d", handle_size);
+		return -1;
+	}
+
+	*handle = k_alloc(handle_size, K_NO_WAIT);
```

Comment:
No this is just an example of how the application could allocate memory for the module. This could be static (well soon it could) or it could be slabs, system heap, shared heaps, user defined heaps or memory blocks. I chose this as it is the most common method in C.